### PR TITLE
initial support for encoding-type

### PR DESCRIPTION
**Pull request by dleute**
This should enable support. I think I caught all the places blobName is put into xml data.

Take a look!

### CODE REVIEW

**Comment by dleute** [comment]
This should fix issue #106.

I have not figured out the best way to test this yet.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 1]
```diff
@@ -148,7 +149,8 @@
             "prefix",
             "Signature",
             "uploadId",
-            "uploads"
+            "uploads",
+            "encoding-type"
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Can you alphabetize these?

**Comment by andrewgaul** [comment]
This fails Travis, can you look at the compilation errors (and likely Checkstyle violations)?

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 2]
```diff
@@ -2280,4 +2280,13 @@ private static String maybeQuoteETag(String eTag) {
     private static boolean startsWithIgnoreCase(String string, String prefix) {
         return string.toLowerCase().startsWith(prefix.toLowerCase());
     }
+
+    // Protects against illegal characters in blob names only if the client requests it
+    private String encodeBlob(String encodingType, String blobName) {
+        if (encodingType.equals("url")) {
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: `HttpServletRequest.getParameter` can return null and thus we must check `encodingType` here.

Comment (dleute) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: .equals in the function does that by default. Supposedly.

Sent from my iPhone

On Dec 11, 2015, at 9:25 PM, Andrew Gaul <notifications@github.com> wrote:

In src/main/java/org/gaul/s3proxy/S3ProxyHandler.java
<https://github.com/andrewgaul/s3proxy/pull/106#discussion_r47425453>:

> @@ -2280,4 +2280,13 @@ private static String maybeQuoteETag(String eTag) {
>      private static boolean startsWithIgnoreCase(String string, String prefix) {
>          return string.toLowerCase().startsWith(prefix.toLowerCase());
>      }
> +
> +    // Protects against illegal characters in blob names only if the client requests it
> +    private String encodeBlob(String encodingType, String blobName) {
> +        if (encodingType.equals("url")) {

HttpServletRequest.getParameter can return null and thus we must check
encodingType here.

—
Reply to this email directly or view it on GitHub
<https://github.com/andrewgaul/s3proxy/pull/106/files#r47425453>.

**Comment by andrewgaul** [comment]
Neither the AWS Java SDK nor jclouds support the url encoding-type presently so we cannot add a unit test.  Could you share a small aws-cli example so I can verify this manually?

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java (hunk 2 above)
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: `Object.equals` checks its parameters against null but if `encodingType` is null then we throw a `NullPointerException` before calling `equals`.

**Comment by dleute** [comment]
With the lates awscli 1.9.12 just do an ls of a bucket. It always sends
encoding type for me.

Sent from my iPhone

On Dec 11, 2015, at 9:27 PM, Andrew Gaul <notifications@github.com> wrote:

Neither the AWS Java SDK nor jclouds support the url encoding-type
presently so we cannot add a unit test. Could you share a small aws-cli
example so I can verify this manually?

—
Reply to this email directly or view it on GitHub
<https://github.com/andrewgaul/s3proxy/pull/106#issuecomment-164101413>.

**Comment by dleute** [comment]
I'll be working on this again on monday.

**Comment by dleute** [comment]
FYI: swift has entirely spoiled me. I don't think about null pointer exceptions anymore.

**Comment by dleute** [comment]
I just pushed in the requested changes. Two comments:

1. I handled the null pointer case in the function since that's where the actual danger is. Not sure if this fits your coding style. I felt it's cleaner than having the same check capability in many parts of the code.

2. I handled the exception, but I'm not sure what you would want the behavior when an unknown encoding is used. I just returned the original string so nothing crashes.It's using only utf-8 so this error should really never happen.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 3]
```diff
@@ -2280,4 +2275,17 @@ private static String maybeQuoteETag(String eTag) {
...
         return string.toLowerCase().startsWith(prefix.toLowerCase());
     }
+
+    // Protects against illegal characters in blob names only if the client requests it
+    private String encodeBlob(String encodingType, String blobName) {
+        if (encodingType != null && encodingType.equals("url")) {
+            try {
+                return URLEncoder.encode(blobName, "UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                return blobName;
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Can you `throw Throwables.propagate(e)` since this exception should never be thrown?

**Comment by andrewgaul** [comment]
Please address Checkstyle violations listed by Travis.

**Comment by andrewgaul** [comment]
Looking at https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html it seems like we should handle `encoding-type` in list objects as well as emit the Encoding Type back to the client.

**Comment by dleute** [comment]
Yeah. We also are supposed to encode those other things: ```Delimiter, Marker, Prefix, NextMarker, Key.```

Right now all we are doing is Key for blobNames.

I imagine we also only include and encode if the encoding-type parameter is "url".

**Comment by dleute** [comment]
Just pushed in the updates based on that document. I tried to match the styles in the function already.

Take a look.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 4]
```diff
@@ -1654,7 +1667,8 @@ private void handleListParts(HttpServletRequest request,
             xml.writeDefaultNamespace(AWS_XMLNS);
 
             writeSimpleElement(xml, "Bucket", containerName);
-            writeSimpleElement(xml, "Key", blobName);
+            writeSimpleElement(xml, "Key", encodeBlob(
+                    request.getParameter("encoding-type"), blobName));
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: `handleListParts` needs an `Encoding-Type` element: https://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadListParts.html

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 5]
```diff
@@ -2280,4 +2294,18 @@ private static String maybeQuoteETag(String eTag) {
     private static boolean startsWithIgnoreCase(String string, String prefix) {
         return string.toLowerCase().startsWith(prefix.toLowerCase());
     }
+
+    // Protects against illegal characters in
+    // blob names only if the client requests it
+    private String encodeBlob(String encodingType, String blobName) {
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Should be `static`.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java (hunk 4 above)
Comment (dleute) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: According to that link, the ancestor is ListBucketResult not ListPartsResult. Not sure if that's a typo, but that isn't in the ListPartsResult area. If you want I can add it and assume it's a typo in the documentation. See Ancestor in attached screenshot. (The ancestor of all the other items is ListPartsResult)

screen shot 2015-12-14 at 4 48 28 pm

**Comment by andrewgaul** [comment]
I looked through the boto source which aws cli uses and it sends `encoding-type` for

* `list` and `get_all_keys`
* `list_versions`
* `list_multipart_uploads` and `get_all_multipart_uploads`
* `get_all_parts`

Based on this, I removed `encoding-type` handling from `handleMultiBlobRemove`, `handleInitiateMultipartUpload`, and `handleCompleteMultipartUpload`.  While I agree that your pull request seems more correct, I would rather err on the side of compatibility with boto which is one of the most popular clients.  I amended your commit, reworded the commit message, and pushed to master as d8c73bc86d30e885e743eb7afcd41cf0b9e9497b.  Thank you for your contribution @dleute!

**Comment by dleute** [comment]
Happy to help!

Sent from my iPhone

On Dec 14, 2015, at 8:13 PM, Andrew Gaul <notifications@github.com> wrote:

I looked through the boto source which aws cli uses and it sends
encoding-type for

   - list and get_all_keys
   - list_versions
   - list_multipart_uploads and get_all_multipart_uploads
   - get_all_parts

Based on this, I removed encoding-type handling from handleMultiBlobRemove,
handleInitiateMultipartUpload, and handleCompleteMultipartUpload. While I
agree that your pull request seems more correct, I would rather err on the
side of compatibility with boto which is one of the most popular clients. I
amended your commit, reworded the commit message, and pushed to master as
d8c73bc
<https://github.com/andrewgaul/s3proxy/commit/d8c73bc86d30e885e743eb7afcd41cf0b9e9497b>.
Thank you for your contribution @dleute <https://github.com/dleute>!

—
Reply to this email directly or view it on GitHub
<https://github.com/andrewgaul/s3proxy/pull/106#issuecomment-164610343>.

**Pull request closed without merging**